### PR TITLE
feat(Gridintersect): new grid intersection options

### LINKED ISSUE #1216: GridIntersect fails with specific shapely Polygon due to change of nrow, ncol in the discretization object
The code below works if ncol = 50 and nrow = 80, but fails if ncol = 25 and nrow = 40. Am still trying to figure out why, but in any case I though it should be reported.

This is the result of the version printing:

3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]
numpy version: 1.20.1
flopy version: 3.3.5
matplotlib version: 3.3.4
shapely version: 1.7.1

**Result with nrow=50, ncol=80**
image

**Error with nrow=25, ncol=40**

KeyError                                  Traceback (most recent call last)
C:\Users\XXXXX~1\AppData\Local\Temp/ipykernel_20404/762291030.py in <module>
     74 
     75 
---> 76 intersection  = gridint.intersect(shply_polygon)
     77 
     78 

~\.conda\envs\flopy-develop\lib\site-packages\flopy\utils\gridintersect.py in intersect(self, shp, **kwargs)
    231                 and self.mfgrid.grid_type == "structured"
    232             ):
--> 233                 rec = self._intersect_polygon_structured(shp)
    234             else:
    235                 rec = self._intersect_polygon_shapely(shp, sort_by_cellid)

~\.conda\envs\flopy-develop\lib\site-packages\flopy\utils\gridintersect.py in _intersect_polygon_structured(self, shp)
   1345                         )
   1346                     else:
-> 1347                         v_realworld = intersect.__geo_interface__[
   1348                             "coordinates"
   1349                         ]

KeyError: 'coordinates'

 
**Code**

```
import sys
import os
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.colors as clt
import flopy

from flopy.utils.gridintersect import GridIntersect
from flopy.utils.geometry import shape
import flopy.discretization as fgrid

import shapely
from shapely.geometry import Point, MultiPoint, LineString, Polygon


print(sys.version)
print('numpy version: {}'.format(np.__version__))
print('flopy version: {}'.format(flopy.__version__))
print('matplotlib version: {}'.format(mpl.__version__))
print('shapely version: {}'.format(shapely.__version__))


points  = [
(-18.261405672010028,493.7361282367448),
(20.0,490.0),
(80.0,490.0),
(120.0,450.0),
(150.0,380.0),
(160.0,350.0),
(160.0,310.0),
(160.0,290.0),
(150.0,280.0),
(140.0,250.0),
(130.0,240.0),
(120.0,230.0),
(90.0,220.0),
(90.0,220.0),
(70.0,240.0),
(60.0,290.0),
(30.0,280.0),
(20.0,250.0),
(20.0,250.0),
(20.0,210.0),
(20.0,170.0),
(10.0,160.0),
(10.0,160.0),
(0.0,150.0),
(-22.404438964241834,169.54377311960536),
(-70.04932182490768,284.51294697903825),
(-79.37114673242925,394.30332922318127),
(-83.51418002466107,444.019728729963)
]

shply_polygon = Polygon(shell=points)

shply_polygon


Lx      = 500.              
Ly      = 800.              

nlay    = 1                
ncol    = 50
nrow    = 80           
             
delr    = np.ones(ncol)*Lx / ncol        
delc    = np.ones(nrow)*Ly / nrow    

sgr = fgrid.StructuredGrid(delc, delr, top=None, botm=None, xoff=0, yoff=0, angrot=0)  


gridint  = GridIntersect(sgr, method="structured", rtree=False)


intersection  = gridint.intersect(shply_polygon)



fig, ax = plt.subplots(1, 1, figsize=(8, 8))

sgr.plot(ax=ax)

gridint.plot_polygon(intersection,ax = ax, lw=2) 
```

### LINKED ISSUE #1467: bug in gridintersect when intersecting linestrings starting in vertex with method="structured"
I came across a bug in the structured GridIntersect routines, when linestrings start in a vertex (not on the model boundary) and continue towards the southeast. The initial intersect calculation to find the initial grid cell uses the starting point of the linestring. This returns the cell with the lowest cellid, which lies to the northwest of the vertex. Then its neighbors (north, south, east, west) are checked for intersection results, which misses the cell to the southeast of the vertex. This means the line cannot be followed through the grid and the intersection result is incomplete. 

I have a fix that works, but doesn't feel very elegant. If the "check neighbors" method does not yield any intersections with length > 0, the southeastern neighboring cell is also checked for an intersection result. This increases the number of intersection calculations slightly for some cases, but does deal with the issue described above.

I will submit as part of the gridintersect PR coming up.   

Code to reproduce:

```python
import matplotlib.pyplot as plt
import numpy as np
from shapely.geometry import LineString, MultiLineString

import flopy.discretization as fgrid
from flopy.utils.gridintersect import GridIntersect

delc = 10 * np.ones(10, dtype=float)
delr = 10 * np.ones(10, dtype=float)

xoff = 0.0
yoff = 0.0
angrot = 0.0
sgr = fgrid.StructuredGrid(
    delc, delr, top=None, botm=None, xoff=xoff, yoff=yoff, angrot=angrot
)

ls1 = LineString([(95, 105), (30, 50)])
ls2 = LineString([(30, 50), (90, 22)])
ls3 = LineString([(90, 22), (0, 0)])
mls = MultiLineString(lines=[ls1, ls2, ls3])

ix = GridIntersect(sgr, method="structured")

r = ix.intersect(mls)

# necessary because ixshapes contains lists of geoms
r["isxhapes"] = [ishp for ishp in r["ixshapes"]]

# note missing portion of intersection result
fig, ax = plt.subplots(1, 1, figsize=(8, 8))
sgr.plot(ax=ax)
ix.plot_linestring(r, ax=ax, cmap="viridis")
plt.show()
```

### PR DESCRIPTION
**Pull request by dbrakenhoff**
Features:
- Add `min_area_fraction` kwarg for polygon intersections, keeping only results that cover a certain fraction of a grid cell (as discussed in #1216)
- Add `contains_centroid` kwarg for polygon intersections, keeping only results that contain cell centroid in intersection result (as discussed in #1216)
- Add `return_all_intersections` kwarg, which return all multiple intersection results for points and linestrings on cell boundaries (as discussed but not quite implemented in #1198 and #1214)

Bug fixes:
- fixes #1467
- fixes #1216 

Other improvements:
- Update example notebook
- Add tests for new features and #1467
- Deal with some shapely DeprecationWarnings

### CODE REVIEW

**Comment by dbrakenhoff** [comment]
I also wanted to deal with the shapely DeprecationWarning "ShapelyDeprecationWarning: Setting custom attributes on geometry objects is deprecated, and will raise an AttributeError in Shapely 2.0".

But I was wondering what the policy is on backward compatibility for optional dependencies? 

The new methods needed to silence this warning are available from shapely 1.8.0 onwards. Do we continue supporting shapely<1.8 for a while, or can we break the code for older versions of shapely?

Let me know, and i will commit a fix accordingly.

File: autotest/t065_test_gridintersect.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
-import os
+import sys
 
+sys.path.insert(1, "..")
```
Comment (jlarsen-usgs) [inline autotest/t065_test_gridintersect.py:3]: Why does the flopy setup directory need to be inserted into the system path?

**Review by jlarsen-usgs — COMMENTED** [review]
The PR looks good overall. The only question I have is about the 

```python
sys.path.insert(1, "..")
```
call in t065_test_gridintersect.py. Is there a support script that is located within that path that is not explicitly installed into python?

File: autotest/t065_test_gridintersect.py (hunk 1 above)
Comment (w-bonelli) [reply autotest/t065_test_gridintersect.py:3]: I can remove this when merging these changes into my own PR. This test script's name is changed (no more `t***`) and the new test functions will need a couple other changes anyways (using `pytest.importorskip` for optional imports, etc)

**Review by jlarsen-usgs — APPROVED** [review]
Approving this PR. @w-bonelli is going to remove the `sys.path.insert()` call from t065 in his upcoming PR.

**Comment by dbrakenhoff** [comment]
Sorry, the insert was just to get the code to use my fork instead of the installed flopy package. Thanks for dealing with that!

Any advice on dealing with that shapely deprecation warning?

**Comment by jlarsen-usgs** [comment]
@dbrakenhoff 

It looks like we won't be able to use `p.name`, which is used to store the cellid, once shapely 2.0.0 is released. Would it break anything or cause performance issues to instead store a dictionary of {name : shapely geometry object} and then update the internals to work with a convention like this instead of storing the cellid directly in the object?

**Comment by dbrakenhoff** [comment]
That would certainly be an option, but not sure if that would hurt performance. 

I'd prefer to go the [method Shapely 2.0](https://shapely.readthedocs.io/en/stable/project.html#a2-2021-07-15) suggests:

```python
geoms, items = self._get_gridshapes()  # modify method to return geoms and cellids
self.strtree = strtree.STRtree(geoms, items)
# then get either geoms or cellids internally with: 
self.strtree.query_geoms()
self.strtree.query_items()
```

Only trouble is then we'd need to add logic for Shapely < 1.8 users if that is something we still want to support. I'd be okay with requiring shapely > 1.8 since it's an optional dependency anyway, but not sure how you guys think about that.

PS. this is probably something for a new issue.

**Comment by langevin-usgs** [comment]
I support requiring shapely > 1.8 unless it wasn't available for one of our supported operating systems.  It looks like win, Mac, and linux are all supported.

**Comment by jdhughes-usgs** [comment]
@dbrakenhoff I would say it would be fine to support `shapely > 1.8` now in develop and the in the next flopy release

**Pull request merged**